### PR TITLE
Sector Nord AG: Added optional check condition in TransitionActions

### PR DESCRIPTION
**Pull request by paulfolkers**
## Proposed change

Additional feature that provides an optional parameter "_Check" in TransitionActions.
Provided by Sector Nord AG as discussed with Roy Kaldung and Henrik Vetter.

## Type of change
1 - 🚀 feature

## Additional information

This parameter is optional.  
To use it, add a key-value-pair to your TransitionAction where key is '_Check' and value is the name of your field to be checked, e.g. 'DynamicField_WorkDone'. It does not check for specific values or patterns, but only if there is any value is set for this field in the affected ticket.

## Checklist

- [x] The code change is tested and works locally.(❗)
- [x] There is no commented out code in this PR.(❕)
- [x] You improved or added new unit tests.(❕)
- [x] Local ZnunyCodePolicy run passes successfully.(❕)
- [x] Local unit tests pass.(❕)
- [x] GitHub workflow ZnunyCodePolicy passes.(❗)
- [x] GitHub workflow unit tests pass.(❗)

### CODE REVIEW

**Comment by rkaldung** [comment]
@paulfolkers Thank you. I probably would like to change the name if the parameter, are you ok with this?

**Comment by paulfolkers** [comment]
> @paulfolkers Thank you. I probably would like to change the name if the parameter, are you ok with this?

@rkaldung Sure, for us it's fine.

**Comment by paulfolkers** [comment]
@rkaldung @dennykorsukewitz Since the label is "wait for contributor", do you need anything else from us here?

**Comment by dennykorsukewitz** [comment]
Hi @paulfolkers 

no, everything's fine so far.
I'll have a look at it soon.

File: Kernel/System/ProcessManagement/TransitionAction/Base.pm [hunk 1]
```diff
@@ -66,6 +66,18 @@ sub _CheckParams {
...
+    if ( $Param{Config}->{_Check} ) {
+        my $DynamicField = $Param{Config}->{_Check};
+        if ( !($Param{Ticket}->{$DynamicField}) ) {
+            return;
+        }
+    }
+    # delete the config entry to make sure it doesn't cause errors
+    if ( exists($Param{Config}->{_Check}) ) {
+        delete($Param{Config}->{_Check});
+    }
```
Comment (dennykorsukewitz) [inline Kernel/System/ProcessManagement/TransitionAction/Base.pm]:
```suggestion
    # Check if we have a needed attribute
    if ( $Param{Config}->{ProcessManagementTransitionCheck} ) {
        my $Attribute = $Param{Config}->{ProcessManagementTransitionCheck};
        return if !$Param{Ticket}->{$Attribute};
    }

    # delete the config entry to make sure it doesn't cause errors
    if ( exists($Param{Config}->{ProcessManagementTransitionCheck}) ) {
        delete($Param{Config}->{ProcessManagementTransitionCheck});
    }
```

File: scripts/test/ProcessManagement/TransitionAction/TransitionActionCheckCondition.t [hunk 2]
```diff
@@ -0,0 +1,716 @@
...
+        },
+    },
+};
+
+my $WebserviceID = $WebserviceObject->WebserviceAdd(
+    Config => $WebserviceConfig,
+    Name => "Test $RandomID",
+    ValidID => 1,
+    UserID => 1,
+);
```
Comment (dennykorsukewitz) [inline scripts/test/ProcessManagement/TransitionAction/TransitionActionCheckCondition.t]:
```suggestion
my $HelperObject       = $Kernel::OM->Get('Kernel::System::UnitTest::Helper');
my $TicketObject       = $Kernel::OM->Get('Kernel::System::Ticket');
my $BackendObject      = $Kernel::OM->Get('Kernel::System::DynamicField::Backend');
my $DynamicFieldObject = $Kernel::OM->Get('Kernel::System::DynamicField');
my $ServiceObject      = $Kernel::OM->Get('Kernel::System::Service');
my $SLAObject          = $Kernel::OM->Get('Kernel::System::SLA');
my $WebserviceObject   = $Kernel::OM->Get('Kernel::System::GenericInterface::Webservice');

$HelperObject->ConfigSettingChange(
    Valid => 1,
    Key => 'Ticket::Type',
    Value => 1,
);

my $RandomID               = $HelperObject->GetRandomID();
my $DFName1                = 'Test1' . $RandomID;
my ( $UserLogin, $UserID ) = $HelperObject->TestUserCreate();
my $TestUserLogin          = $HelperObject->TestCustomerUserCreate();

my $TicketID = $TicketObject->TicketCreate(
    Title         => 'test',
    QueueID       => 1,
    Lock          => 'unlock',
    Priority      => '3 normal',
    StateID       => 1,
    TypeID        => 1,
    OwnerID       => 1,
    ResponsibleID => 1,
    UserID        => $UserID,
);

my $Success = $TicketObject->TicketCustomerSet(
    User     => $TestUserLogin,
    TicketID => $TicketID,
    UserID   => 1,
);

my %DynamicFieldConfig = (
    Name       => $DFName1,
    Label      => $DFName1,
    FieldType  => 'Dropdown',
    ObjectType => 'Ticket',
    Config     => {
        TranslatableValues => '0',
        PossibleValues => {
            1 => 'A',
            2 => 'B',
            3 => 'C',
        },
    },
);

my $ID = $DynamicFieldObject->DynamicFieldAdd(
    %DynamicFieldConfig,
    FieldOrder => 99999,
    ValidID    => 1,
    UserID     => 1,
);

my $DynamicField = $DynamicFieldObject->DynamicFieldGet(
    ID => $ID,
);

$Self->True(
    $ID,
    "DynamicFieldCreated",
);

# add service
my $ServiceID = $ServiceObject->ServiceAdd(
    Name    => 'UnitTest'. $RandomID,
    ValidID => 1,
    UserID  => 1,
);

$ServiceObject->CustomerUserServiceMemberAdd(
    CustomerUserLogin => $TestUserLogin,
    ServiceID         => $ServiceID,
    Active            => 1,
    UserID            => 1,
);

my %ServiceData = $ServiceObject->ServiceGet(
    ServiceID => $ServiceID,
    UserID    => 1,
);

# add SLA
my $SLAID = $SLAObject->SLAAdd(
    ServiceIDs => [$ServiceID],
    Name       => 'SLA Name'. $RandomID,
    Calendar   => 'Calendar1',
    ValidID    => 1,
    UserID     => 1,
);

# web service config
my $WebserviceConfig = {
    Debugger => {
        DebugThreshold => 'debug',
        TestMode       => 1,
    },
    Requester => {
        Transport => {
            Type   => 'HTTP::Test',
            Config => {
                Fail => 0,
            },
        },
        Invoker => {
            test_operation => {
                Type => 'Test::TestSimple',
            },
        },
    },
};

my $WebserviceID = $WebserviceObject->WebserviceAdd(
    Config  => $WebserviceConfig,
    Name    => "Test $RandomID",
    ValidID => 1,
    UserID  => 1,
);
```

Comment (dennykorsukewitz) [inline scripts/test/ProcessManagement/TransitionAction/TransitionActionCheckCondition.t]:
```suggestion
my $HelperObject       = $Kernel::OM->Get('Kernel::System::UnitTest::Helper');
my $TicketObject       = $Kernel::OM->Get('Kernel::System::Ticket');
my $BackendObject      = $Kernel::OM->Get('Kernel::System::DynamicField::Backend');
my $DynamicFieldObject = $Kernel::OM->Get('Kernel::System::DynamicField');
my $ServiceObject      = $Kernel::OM->Get('Kernel::System::Service');
my $SLAObject          = $Kernel::OM->Get('Kernel::System::SLA');
my $WebserviceObject   = $Kernel::OM->Get('Kernel::System::GenericInterface::Webservice');

$HelperObject->ConfigSettingChange(
    Valid => 1,
    Key => 'Ticket::Type',
    Value => 1,
);

my $RandomID               = $HelperObject->GetRandomID();
my $DFName1                = 'Test1' . $RandomID;
my ( $UserLogin, $UserID ) = $HelperObject->TestUserCreate();
my $TestUserLogin          = $HelperObject->TestCustomerUserCreate();

my $TicketID = $TicketObject->TicketCreate(
    Title         => 'test',
    QueueID       => 1,
    Lock          => 'unlock',
    Priority      => '3 normal',
    StateID       => 1,
    TypeID        => 1,
    OwnerID       => 1,
    ResponsibleID => 1,
    UserID        => $UserID,
);

my $Success = $TicketObject->TicketCustomerSet(
    User     => $TestUserLogin,
    TicketID => $TicketID,
    UserID   => 1,
);

my %DynamicFieldConfig = (
    Name       => $DFName1,
    Label      => $DFName1,
    FieldType  => 'Dropdown',
    ObjectType => 'Ticket',
    Config     => {
        TranslatableValues => '0',
        PossibleValues => {
            1 => 'A',
            2 => 'B',
            3 => 'C',
        },
    },
);

my $ID = $DynamicFieldObject->DynamicFieldAdd(
    %DynamicFieldConfig,
    FieldOrder => 99999,
    ValidID    => 1,
    UserID     => 1,
);

my $DynamicField = $DynamicFieldObject->DynamicFieldGet(
    ID => $ID,
);

$Self->True(
    $ID,
    "DynamicFieldCreated",
);

# add service
my $ServiceID = $ServiceObject->ServiceAdd(
    Name    => 'UnitTest'. $RandomID,
    ValidID => 1,
    UserID  => 1,
);

$ServiceObject->CustomerUserServiceMemberAdd(
    CustomerUserLogin => $TestUserLogin,
    ServiceID         => $ServiceID,
    Active            => 1,
    UserID            => 1,
);

my %ServiceData = $ServiceObject->ServiceGet(
    ServiceID => $ServiceID,
    UserID    => 1,
);

# add SLA
my $SLAID = $SLAObject->SLAAdd(
    ServiceIDs => [$ServiceID],
    Name       => 'SLA Name'. $RandomID,
    Calendar   => 'Calendar1',
    ValidID    => 1,
    UserID     => 1,
);

# web service config
my $WebserviceConfig = {
    Debugger => {
        DebugThreshold => 'debug',
        TestMode       => 1,
    },
    Requester => {
        Transport => {
            Type   => 'HTTP::Test',
            Config => {
                Fail => 0,
            },
        },
        Invoker => {
            test_operation => {
                Type => 'Test::TestSimple',
            },
        },
    },
};

my $WebserviceID = $WebserviceObject->WebserviceAdd(
    Config  => $WebserviceConfig,
    Name    => "Test $RandomID",
    ValidID => 1,
    UserID  => 1,
);
```

File: scripts/test/ProcessManagement/TransitionAction/TransitionActionCheckCondition.t [hunk 3]
```diff
@@ -0,0 +1,716 @@
+# --
```
Comment (dennykorsukewitz) [inline scripts/test/ProcessManagement/TransitionAction/TransitionActionCheckCondition.t]: Please move this test to `scripts/test/ProcessManagement/TransitionAction/Base/TransitionActionCheckCondition.t`

Unfortunately, github doesn't let me edit so many lines at once and each change individually would be just work for both of us.
So I have uploaded the test for you here:

[TransitionActionCheckCondition.t.zip](https://github.com/znuny/Znuny/files/7285727/TransitionActionCheckCondition.t.zip)

**Review by dennykorsukewitz — CHANGES_REQUESTED** [review]
Hi @paulfolkers,

Good work so far.
I have changed a few things.
The test is great. :+1:.
I have only adapted the Znuny CodeStyle.

🚀  Thank you.

File: scripts/test/ProcessManagement/TransitionAction/TransitionActionCheckCondition.t (hunk 3 above)
Comment (paulfolkers) [reply scripts/test/ProcessManagement/TransitionAction/TransitionActionCheckCondition.t]: I committed your uploaded file, should be part of this PR now.
Thanks for your work & feedback!

File: scripts/test/ProcessManagement/TransitionAction/TransitionActionCheckCondition.t [hunk 4]
```diff
@@ -0,0 +1,715 @@
...
+# Copyright (C) 2021 Znuny GmbH, https://znuny.org/
+# --
+# This software comes with ABSOLUTELY NO WARRANTY. For details, see
+# the enclosed file COPYING for license information (GPL). If you
+# did not receive this file, see https://www.gnu.org/licenses/gpl-3.0.txt.
```
Comment (dennykorsukewitz) [inline scripts/test/ProcessManagement/TransitionAction/TransitionActionCheckCondition.t]: Hi @paulfolkers 

sorry for that:

```suggestion
# This software comes with ABSOLUTELY NO WARRANTY. For details, see
# the enclosed file COPYING for license information (AGPL). If you
# did not receive this file, see http://www.gnu.org/licenses/agpl.txt.
```

Please move this test to scripts/test/ProcessManagement/TransitionAction/Base/TransitionActionCheckCondition.t

**Review by dennykorsukewitz — CHANGES_REQUESTED** [review]

File: scripts/test/ProcessManagement/TransitionAction/TransitionActionCheckCondition.t (hunk 4 above)
Comment (paulfolkers) [reply scripts/test/ProcessManagement/TransitionAction/TransitionActionCheckCondition.t]: No problem, committed. ;-)

**Comment by paulfolkers** [comment]
Sorry, #2 and #3 should not have been mentioned in commits.

**Comment by dennykorsukewitz** [comment]
Hi @paulfolkers 

> Sorry, #2 and #3 should not have been mentioned in commits.

No problem 😄 

- [ ] Please move this UnitTest to scripts/test/ProcessManagement/TransitionAction/`Base`/TransitionActionCheckCondition.t

**Review by dennykorsukewitz — APPROVED** [review]

**Comment by paulfolkers** [comment]
@dennykorsukewitz Committed the changed path. ✔

**Review by dennykorsukewitz — APPROVED** [review]

**Pull request merged**